### DISCUSSION TITLE
numpy to PyTorch tensor conversion - memory sharing with respect to '+=' and '=' operators

### DISCUSSION THREAD

**Discussion by AiKoushik** [mrdbourke/pytorch-deep-learning#755] [General]
I was going through numpy to tensor and tensor to numpy conversion, where it was mentioned that memory is not shared between the original and the new tensor/array.

But while performing `+=` for increment, they behaved like they are sharing the same memory. Although the moment I use statements like `tensor = tensor + 1` that sharing stops e.g. `tensor += 1`

Can someone tell me the right topic to read so it doesn't feel like a quirky behaviour?

I am using Google Colab and I'm on CPU.
Code:

```import torch
import numpy as np

array = np.arange(1, 8)
tensor = torch.from_numpy(array)

array, tensor

array += 1
array, tensor

array = array + 1
array, tensor

array += 1
array, tensor

**Comment by sametucuncu** [comment]
The "+=" operates in place, so it will change the original numpy array.
But when you use:
array = array + 1
It is same as:
array2 = array + 1
So basically, you are creating a new variable with a different memory address.

You can try these codes and observe how each variable memory address changes.
(Try them in seperate cells.)

```
import numpy as np
array = np.arange(1, 8)
tensor = torch.from_numpy(array)
# tensor and array share the same memory.

print(f"array: {array}, tensor: {tensor}")
print(f"address of array: {id(array)}, address of tensor: {id(tensor)}")
```

```
array += 1 
print(f"array: {array}, tensor: {tensor}") 
print(f"address of array: {id(array)}, address of tensor: {id(tensor)}")
# You will see the id of array same as above.
```

```
array = array + 1 
print(f"array: {array}, tensor: {tensor}") 
print(f"address of array: {id(array)}, address of tensor: {id(tensor)}")
# You will see the id of array has changed. Because it is different variable.
```

```
array += 1 
print(f"array: {array}, tensor: {tensor}") 
print(f"address of array: {id(array)}, address of tensor: {id(tensor)}")
```

**Reply by AiKoushik** [reply]
Thank you. I understood the same from my searching.
So basically the original tensor and the derived ones share the same memory. It's just a mistake in the related video where it says that these use different memory locations.

**Reply by shisirkha** [reply]
Good explanation of this concept.

I was just scrolling through the discussion page to help others but I'm pretty impressed by your explanation.

Nice.
